fractional-NFT/project/token-distributor-contract/src/interface.sw

```diff
@@ -0,0 +1,162 @@
+library interface;
+
+dep data_structures;
+
+use data_structures::TokenDistribution;
+
+abi TokenDistributor {
+    /// Allows for the NFT owner to start a buyback of tokens at a set price.
+    ///
+    /// # Arguments
+    ///
+    /// * `fractional_nft` - The token contract that holds the NFT.
+    /// * `token_price` - The price for a single fractionalized NFT token.
+    ///
+    /// # Reverts
+    ///
+    /// * When `fractional_nft` does not map to an exising token distribution.
+    /// * When the sender is not the NFT owner.
+    /// * When the distribution is not in the distributing state.
+    /// * When providing an incorrect payment asset or amount.
+    #[storage(read, write)]
+    fn buyback(fractional_nft: ContractId, token_price: u64);
+
+    /// Starts a new token distribution and takes control of the NFT.
+    ///
+    /// # Arguments
+    ///
+    /// * `external_asset` - The asset which will be accepted in return for fractionalized NFT tokens.
+    /// * `fractional_nft` - The token contract that holds the NFT.
+    /// * `nft` - The contract that manages the NFT held by the fractionalized NFT.
+    /// * `owner` - The user which will have the ability to withdraw both the paid tokens and the NFT.
+    /// * `reserve_price` - The price at which ownership of the NFT may be bought outright.
```

Comment:
hm, this allows the owner to sell the entire NFT at a cheaper price than the sum of the parts, likewise it can be more expensive but idk why anyone would do that. Have you considered if this can have side effects in the code somewhere?

Reply:
Coming back to these comments, yes both of these cases are possible but I don't believe this is something we can enforce. Prices can fluctuate for the NFT on the market, the tokens on the market, and the token price can be changed by the admin. What is being sold here is the admin rights and not the NFT itself. 

Within this contract itself there are no side effects.